File: class-oop-implementation/src/scene/SpotLight.java
package scene;

import java.awt.Color;
import math.Intersection;
import math.Ray;
import math.Vector3D;

public class SpotLight implements Light {
  private static final float MIN_DISTANCE_SQ = 1e-4f;
  private static final float MIN_NORM_SQ = 1e-8f;
  private final Vector3D pos;
  private final Vector3D dir;
  private final Vector3D color;
  private final float intensity;
  private final float cosAngle;
  private final float radius;
  private final int samples;

  public SpotLight(Vector3D pos, Vector3D dir, Color color, float intensity, float angleDeg) {
    this(pos, dir, color, intensity, angleDeg, 0.0f, 1);
  }

  public SpotLight(Vector3D pos, Vector3D dir, Color color, float intensity, float angleDeg, float radius, int samples) {
    this.pos = pos;
    this.dir = dir.normalize();
    this.color = Material.fromColor(color);
    this.intensity = intensity;
    this.cosAngle = (float) Math.cos(Math.toRadians(angleDeg / 2.0));
    this.radius = radius;
    this.samples = samples;
  }

  public SpotLight(Vector3D pos, Vector3D dir, Vector3D colorVec, float intensity, float angleDeg, float radius, int samples) {
    this.pos = pos;
    this.dir = dir.normalize();
    this.color = colorVec;
    this.intensity = intensity;
    this.cosAngle = (float) Math.cos(Math.toRadians(angleDeg / 2.0));
    this.radius = radius;
    this.samples = samples;
  }

  private static Vector3D safeNormalize(Vector3D v) {
    if (v.lengthSquared() <= MIN_NORM_SQ) {
      return new Vector3D(0f, 0f, 0f);
    }
    return v.normalize();
  }

  @Override
  public Vector3D shade(Material material, Intersection hit, Vector3D viewDir) {
    final Vector3D toLight = pos.sub(hit.getPoint());
    final float distSq = Math.max(MIN_DISTANCE_SQ, toLight.lengthSquared());
    final Vector3D n = safeNormalize(hit.getNormal());
    final Vector3D l = safeNormalize(toLight);
    final Vector3D v = safeNormalize(viewDir);

    final float cosTheta = l.mul(-1f).dot(dir);
    float spotAttenuation = 0.0f;
    if (cosTheta >= cosAngle) {
      spotAttenuation = 1.0f;
    }

    if (spotAttenuation <= 0.0f) {
      return new Vector3D(0f, 0f, 0f);
    }

    final float lambert = Math.max(0f, n.dot(l));
    final Vector3D reflection = safeNormalize(l.mul(-1f).sub(n.mul(2f * l.mul(-1f).dot(n))));
    final float specBase = Math.max(0f, v.dot(reflection));
    final float specular = (float) Math.pow(specBase, Math.max(1f, material.getShininess()));
    final Vector3D radiance = color.mul((intensity / distSq) * spotAttenuation);
    return radiance.vec_mul(material.getDiffuse().mul(lambert).add(material.getSpecular().mul(specular)));
  }

  @Override
  public ShadowRay shadowRay(Intersection hit) {
    Vector3D toLight = pos.sub(hit.getPoint());
    float distance = (float) Math.sqrt(toLight.lengthSquared());
    Vector3D origin = hit.getPoint().add(hit.getNormal().normalize().mul(Light.SHADOW_BIAS));
    return new ShadowRay(new Ray(origin, toLight), distance);
  }

  public Vector3D getPos() {
    return pos;
  }

  public Vector3D getDir() {
    return dir;
  }

  public Vector3D getColor() {
    return color;
  }

  public float getIntensity() {
    return intensity;
  }

  public float getCosAngle() {
    return cosAngle;
  }

  public float getRadius() {
    return radius;
  }

  public int getSamples() {
    return samples;
  }
}

File: src/spot_light.cpp
module spot_light;
import color_rgb;
import glm;
import interval;
import ray;
import hitevent;
import material;

namespace cg
{
[[nodiscard]] color_rgb shade_phong(const material& m, const hitevent& he,
                                    const spot_light& l, vec3 view_dir)
{
    constexpr static double kMinDistanceSq = 1e-7f;

    const vec3 to_light = l.pos - he.p;
    const double dist_sq =
        glm::max(kMinDistanceSq, glm::dot(to_light, to_light));
    const double dist = glm::sqrt(dist_sq);
    const vec3 light_dir = to_light / dist;

    // Spotlight direction attenuation
    const double cos_theta = glm::dot(-light_dir, l.dir);
    double spot_attenuation = 0.0;
    if (cos_theta >= l.cos_angle)
    {
        spot_attenuation = 1.0;
    }

    if (spot_attenuation <= 0.0)
        return color_rgb{0.0, 0.0, 0.0};

    const color_rgb radiance = color_rgb{l.color * (l.intensity / dist_sq) * spot_attenuation};
    return phong_brdf(m, he, to_light, radiance, view_dir);
}

auto get_shadow_info(const spot_light& light, const hitevent& hit,
                     const double kShadowBias) -> std::pair<ray, interval>
{
    const vec3 to_light = light.pos - hit.p;
    const double light_distance = glm::length(to_light);
    const vec3 shadow_origin = hit.p + hit.normal * kShadowBias;
    const ray shadow_ray{shadow_origin, to_light};
    if (light_distance <= kShadowBias)
        return {shadow_ray, empty_interval};
    const interval shadow_range{kShadowBias, light_distance - kShadowBias};
    return {shadow_ray, shadow_range};
}
} // namespace cg
